sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala

```diff
@@ -1304,6 +1304,15 @@ class DatasetSuite extends QueryTest with SharedSQLContext {
       assert(rlike3.count() == 0)
     }
   }
+
+  test("SPARK-21538: Attribute resolution inconsistency in Dataset API") {
+    val df = spark.range(1).withColumnRenamed("id", "x")
+    checkAnswer(df.sort(col("id")), df.sort("id"))
+    checkAnswer(df.sort($"id"), df.sort("id"))
+    checkAnswer(df.sort('id), df.sort("id"))
+    checkAnswer(df.orderBy('id), df.sort("id"))
+    checkAnswer(df.orderBy("id"), df.sort("id"))
```

Comment:
Indeed, looks much better. I appreciate the explanation and will take this into account in the future. I will update the test in a minute, thanks